physical/raft/fsm.go

```diff
@@ -243,6 +251,90 @@ func writeSnapshotMetaToDB(metadata *raft.SnapshotMeta, db *bolt.DB) error {
 	return nil
 }
 
+func (f *FSM) upgradeConfiguration() error {
+
+	config := f.latestConfig.Load().(*ConfigurationValue)
+
+	// When a fresh cluster is being brought up, there won't be a config. Do
+	// nothing in that case.
+	if config == nil {
+		return nil
+	}
+
+	upgraded := false
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			// DesiredSuffrage not being set is the upgrade criterion. Use the
+			// current suffrage of the node in the raft cluster as the desired
+			// suffrage, as there is no other better option.
+			if srv.DesiredSuffrage == "" {
+				switch srv.Suffrage {
+				case int32(raft.Nonvoter):
+					srv.DesiredSuffrage = "non-voter"
+				default:
+					srv.DesiredSuffrage = "voter"
+				}
+				upgraded = true
+			}
+			// Bring the intent to the fsm instance. This will be used by
+			// autopilot to keep the non-voters from getting promoted.
+			f.desiredSuffrage = srv.DesiredSuffrage
+		}
+	}
+	if upgraded {
+		if err := f.persistConfig(config); err != nil {
+			return err
+		}
+		f.latestConfig.Store(config)
+	}
+	return nil
+}
+
+// This is called when a node successfully joins the cluster. This intent should
+// land in the stored configuration. If the config isn't available yet, we still
+// go ahead and store the intent in the fsm. During the next updated to the
+// configuration, this intent will be persisted.
+func (f *FSM) witnessSuffrage(desiredSuffrage string) error {
+	config := f.latestConfig.Load().(*ConfigurationValue)
+	if config == nil {
+		// This will be persisted when the fsm receives the `LogConfiguration`
+		// log.
+		// TODO: Do we want to explicitly create the config entry here?
+		f.desiredSuffrage = desiredSuffrage
+		return nil
+	}
+
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			srv.DesiredSuffrage = desiredSuffrage
+		}
+	}
+
+	if err := f.persistConfig(config); err != nil {
+		return err
+	}
+
+	f.latestConfig.Store(config)
+	f.desiredSuffrage = desiredSuffrage
+
+	return nil
+}
+
+func (f *FSM) persistConfig(config *ConfigurationValue) error {
+	configBytes, err := proto.Marshal(config)
+	if err != nil {
+		return err
+	}
+
+	return f.db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(configBucketName)
```

Comment:
You should be able to assume the bucket already exists, since openDBFile is called by NewFSM and it already did this call.

Reply:
Yep. Thanks! done.